docs/source/torch_cuda_memory.rst

```diff
@@ -0,0 +1,61 @@
+.. _torch_cuda_memory:
+
+Understanding CUDA Memory Usage
+===============================
+To debug CUDA memory use, PyTorch provides a way to generate memory snapshots that record the state of allocated CUDA memory
+at any point in time, and optionally record the history of allocation events that led up to that snapshot.
+
+The generated snapshots can then be drag and dropped onto the interactiver viewer hosted at `pytorch.org/memory_visualizer <pytorch.org/memory_visualizer>`_ which
+can be used to explore the snapshot.
+
+Generating a Snapshot
+=====================
+The common pattern for recording a snapshot is to enable memory history, run the code to be observed, and then save a file with a pickled snapshot:
+
+.. code-block:: python
+
+   # enable memory history, which will
+   # add tracebacks and event history to snapshots
+   torch.cuda.memory._record_memory_history()
+
+   run_your_code()
+   torch.cuda.memory._dump_snapshot("my_snapshot.pickle")
+
+Using the visualizer
+====================
+
+Open `pytorch.org/memory_visualizer <pytorch.org/memory_visualizer>`_ and drag/drop the pickled snapshot file into the visualizer.
+The visualizer is a javascript application that runs locally on your computer. It does not upload any snapshot data.
```

Comment:
Can we have a simple `python -m http.serve ...` command that people can use if they want to host the visualizer locally?

Reply:
visualizer.html refers to MemoryViz.js refers to d3.js. It's possible but somewhat annoying to keep a packed javascript version of all of that, but I don't know where we would host it. It's also possible to have them host just visualizer.html, but I'd argue that doesn't solve any concerns because the rest is hosted by us. So I didn't do any of these things. Ideally we'd figure out some "webarchive" style thing where you can just open the archive which contains the snapshot and all the visualizer code, but I don't think any of the web archive formats can actually do that.

Reply:
Ho right, that sounds good. Let's just have the online version then I agree.